File: Problem-0001-Two-Sum/Problem-0001-Clean/java/Solution.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

//=================================-Solution-=================================
public class Solution {
    //------------------------------Two-Sum-----------------------------------
    public int[] twoSum(int[] nums, int target) {
        ValueIndexMap valueMap = new ValueIndexMap(nums, target);
        return valueMap.getSumIndexes();
    }
}
//=============================-Value-Index-Map-==============================
class ValueIndexMap {
    public int[] nums;
    public int target;
    public Map<Integer, List<Integer>> valueIndexMap;
    //===========================-Constructors-===============================
    public ValueIndexMap(int[] nums, int target) {
        this.nums = nums;
        this.target = target;
        this.initializeMap();
    }
    //=============================-Methods-==================================

    //---------------------------Initialize-Map-------------------------------
    private void initializeMap() {
       this.valueIndexMap = new HashMap<>();
        for (int index = 0; index < this.nums.length; index++) {
            int valueAtIndex = this.nums[index];
            this.addNumToMap(valueAtIndex, index);
        }
    }

    //--------------------------Get-Sum-Indexes-------------------------------
    public int[] getSumIndexes() {
        for (Map.Entry<Integer, List<Integer>> entry : this.valueIndexMap.entrySet()) {
            int primaryNum = entry.getKey();
            int remainderNum = this.getRemainderNumber(primaryNum);
            if (this.containsValue(remainderNum)) {
                if (primaryNum == remainderNum) {
                    return this.getValueIndexes(primaryNum);
                } else {
                    return this.getValueIndexes(primaryNum, remainderNum);
                }
            }
        }
        return null;
    }
    //-------------------------Get-Value-Indexes------------------------------
    private int[] getValueIndexes(int value) {
        List<Integer> valueIndexes = this.valueIndexMap.get(value);
        int primaryNum = valueIndexes.get(0);
        int secondaryNum = valueIndexes.get(1);
        return new int[] {primaryNum, secondaryNum};
    }
    //-------------------------Get-Value-Indexes------------------------------
    private int[] getValueIndexes(int primaryValue, int secondaryValue) {
        List<Integer> primaryValueIndexes = this.valueIndexMap.get(primaryValue);
        List<Integer> secondaryValueIndexes = this.valueIndexMap.get(secondaryValue);
        int primaryNum = primaryValueIndexes.getFirst();
        int secondaryNum = secondaryValueIndexes.getFirst();
        return new int[] {primaryNum, secondaryNum};
    }
    //-------------------------Add-Numbers-To-Map-----------------------------
    private void addNumToMap(int value, int index) {
        if (!this.containsValue(value)) {
            this.valueIndexMap.put(value, new ArrayList<>());
        }
        this.valueIndexMap.get(value).add(index);
    }
    //---------------------------Contains-Value-------------------------------
    private boolean containsValue(int value) {
        return this.valueIndexMap.containsKey(value);
    }
    //------------------------Get-Remaining-Number----------------------------
    private int getRemainderNumber(int primaryNumber) {
        return this.target - primaryNumber;
    }
}

File: Problem-0001-Two-Sum/Problem-0001-Clean/typescript/solution.ts
//=================================-Solution-=================================

//==================================-Types-===================================
type ValueIndexesMap = Map<number, Array<number>>;

//----------------------------------Two-Sum-----------------------------------
function twoSum(nums: number[], target: number): number[] {
    const map: ValueIndexesMap = new Map();
    for (let index: number = 0; index < nums.length; index++) {
        const value: number = nums[index];
        insertValueIndex(map, value, index);
    }
    return getSumIndexes(map, target);
}
//-----------------------------Insert-Value-Index-----------------------------
function insertValueIndex(map: Map<number, Array<number>>,
                          value: number,
                          index: number): void {
    if (!map.has(value)) {
        map.set(value, new Array<number>());
    }
    const valueArray: Array<number> = map.get(value);
    valueArray.push(index)
    map.set(value, valueArray);
}
//-----------------------------Get-Remainder-Num------------------------------
function getRemainderNum(target: number, value: number): number {
    return target - value;
}
//------------------------------Get-Sum-Indexes-------------------------------
function getSumIndexes(map: ValueIndexesMap, target: number): number[] {
    for (const [value, indexes] of map.entries()) {
        const remainderNum: number = getRemainderNum(target, value);
        if (map.has(remainderNum)) {
            switch (value) {
                case remainderNum:
                    if (indexes.length >= 2) {
                        return [indexes[0], indexes[1]];
                    }
                    break;
                default:
                    const secondaryIndexArray: number[] = map.get(remainderNum);
                    return [indexes[0], secondaryIndexArray[0]];
            }
        }
    }
    return null;
}